File: src/main/java/de/cplaiz/activecraftcore/manager/BanManager.java
package de.cplaiz.activecraftcore.manager;

import de.cplaiz.activecraftcore.ActiveCraftCore;
import de.cplaiz.activecraftcore.events.PlayerBanEvent;
import de.cplaiz.activecraftcore.events.PlayerIpBanEvent;
import de.cplaiz.activecraftcore.events.PlayerUnbanEvent;
import de.cplaiz.activecraftcore.messagesv2.MessageFormatter;
import de.cplaiz.activecraftcore.messagesv2.MessageSupplier;
import de.cplaiz.activecraftcore.playermanagement.Profilev2;
import de.cplaiz.activecraftcore.utils.StringUtils;
import de.cplaiz.activecraftcore.utils.TimeUtils;
import org.bukkit.BanEntry;
import org.bukkit.BanList;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;

import java.util.Date;

public class BanManager {

    public static class IP {
        public static boolean isBanned(String target) {
            for (BanEntry banEntry : Bukkit.getBanList(BanList.Type.IP).getBanEntries()) {
                if (target.equals(banEntry.getTarget())) return true;
            }
            return false;
        }

        public static void ban(String target, String reason, Date expires, String source) {
            Bukkit.getScheduler().runTask(ActiveCraftCore.getInstance(), () -> {
                PlayerIpBanEvent event = new PlayerIpBanEvent(target, true, reason, expires, source);
                Bukkit.getPluginManager().callEvent(event);
                if (!event.isCancelled()) {
                    Bukkit.getBanList(BanList.Type.IP).addBan(target, reason, expires, source);
                    for (Player player : Bukkit.getOnlinePlayers()) {
                        if (player.getAddress().getAddress().toString().replace("/", "").equals(target)) {
                            Profilev2 profile = Profilev2.of(player);
                            MessageSupplier messageSupplier = profile.getMessageSupplier(ActiveCraftCore.getInstance());
                            if (profile != null)
                                profile.setTimesIpBanned(profile.getTimesIpBanned() + 1);
                            Bukkit.getScheduler().runTask(ActiveCraftCore.getInstance(), () -> {
                                String expirationString =
                                        TimeUtils.getRemainingAsString(expires).equalsIgnoreCase("never") ?
                                                messageSupplier.getMessage("command.banscreen.expiration-permanent") :
                                                messageSupplier.getFormatted("command.banscreen.expiration",
                                                        new MessageFormatter(messageSupplier.getActiveCraftMessage(), "time", TimeUtils.getRemainingAsString(expires)));
                                player.kickPlayer(messageSupplier.getMessage("command.banscreen.ip-title", ChatColor.RED)
                                        + "\n \n" + expirationString
                                        + "\n" + messageSupplier.getFormatted("command.banscreen.reason",
                                        new MessageFormatter(messageSupplier.getActiveCraftMessage(), "reason", reason)));
                            });
                        }
                    }
                }
            });
        }

        public static void ban(Player target, String reason, Date expires, String source) {
            ban(target.getName(), reason, expires, source);
        }

        public static void unban(String target) {
            if (!StringUtils.isValidInet4Address(target)) {
                return;
            }
            BanEntry entry = Bukkit.getBanList(BanList.Type.IP).getBanEntry(target);
            if (entry == null) return;
            PlayerIpBanEvent event = new PlayerIpBanEvent(target, false, entry.getReason(), entry.getExpiration(), entry.getSource());
            Bukkit.getPluginManager().callEvent(event);
            if (event.isCancelled()) return;
            Bukkit.getBanList(BanList.Type.IP).pardon(target);
        }

        public static void unban(Player target) {
            unban(target.getName());
        }

        public static java.util.Set<org.bukkit.BanEntry> getBans() {
            return Bukkit.getBanList(BanList.Type.IP).getBanEntries();
        }

        public static BanEntry getBanEntry(String target) {
            return Bukkit.getBanList(BanList.Type.IP).getBanEntry(target);
        }

        public static BanEntry getBanEntry(Player target) {
            return Bukkit.getBanList(BanList.Type.IP).getBanEntry(target.getName());
        }
    }

    public static class Name {
        public static boolean isBanned(String target) {
            return Bukkit.getBanList(BanList.Type.NAME).isBanned(target);
        }

        public static void ban(String target, String reason, Date expires, String source) {
            Bukkit.getScheduler().runTask(ActiveCraftCore.getInstance(), () -> {
                PlayerBanEvent event = new PlayerBanEvent(target, reason, expires, source);
                Bukkit.getPluginManager().callEvent(event);
                if (!event.isCancelled()) {
                    Bukkit.getBanList(BanList.Type.NAME).addBan(target, reason, expires, source);
                    Profilev2 profile = Profilev2.of(target);
                    MessageSupplier messageSupplier = profile.getMessageSupplier(ActiveCraftCore.getInstance());
                    if (profile != null)
                        profile.setTimesBanned(profile.getTimesBanned() + 1);
                    if (Bukkit.getPlayer(target) != null)
                        Bukkit.getScheduler().runTask(ActiveCraftCore.getInstance(), () -> {
                            String expirationString =
                                    TimeUtils.getRemainingAsString(expires).equalsIgnoreCase("never") ?
                                            messageSupplier.getMessage("command.banscreen.expiration-permanent") :
                                            messageSupplier.getFormatted("command.banscreen.expiration",
                                                    new MessageFormatter(messageSupplier.getActiveCraftMessage(), "time", TimeUtils.getRemainingAsString(expires)));
                            Bukkit.getPlayer(target).kickPlayer(messageSupplier.getMessage("command.banscreen.title")
                                    + "\n \n" + expirationString
                                    + "\n" + messageSupplier.getFormatted("command.banscreen.reason",
                                    new MessageFormatter(messageSupplier.getActiveCraftMessage(), "reason", reason)));
                        });
                }
            });
        }

        public static void ban(Player target, String reason, Date expires, String source) {
            ban(target.getName(), reason, expires, source);
        }

        public static void unban(String target) {
            PlayerUnbanEvent event = new PlayerUnbanEvent(target);
            Bukkit.getPluginManager().callEvent(event);
            if (event.isCancelled()) return;
            Bukkit.getBanList(BanList.Type.NAME).pardon(target);
        }

        public static void unban(Player target) {
            unban(target.getName());
        }

        public static java.util.Set<org.bukkit.BanEntry> getBans() {
            return Bukkit.getBanList(BanList.Type.NAME).getBanEntries();
        }

        public static BanEntry getBanEntry(String target) {
            return Bukkit.getBanList(BanList.Type.NAME).getBanEntry(target);
        }

        public static BanEntry getBanEntry(Player target) {
            return Bukkit.getBanList(BanList.Type.NAME).getBanEntry(target.getName());
        }
    }
}

File: src/main/java/org/activecraft/activecraftcore/manager/BanManager.kt
package org.activecraft.activecraftcore.manager

import org.activecraft.activecraftcore.ActiveCraftCore
import org.activecraft.activecraftcore.events.PlayerBanEvent
import org.activecraft.activecraftcore.events.PlayerIpBanEvent
import org.activecraft.activecraftcore.events.PlayerUnbanEvent
import org.activecraft.activecraftcore.messages.MessageFormatter
import org.activecraft.activecraftcore.messages.MessageSupplier
import org.activecraft.activecraftcore.playermanagement.Profile
import org.activecraft.activecraftcore.utils.TimeUtils.getRemainingAsString
import org.activecraft.activecraftcore.utils.isValidInet4Address
import org.bukkit.BanEntry
import org.bukkit.BanList
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.entity.Player
import java.util.*

sealed interface BanManager {

    object IP {
        fun isBanned(target: String): Boolean {
            for (banEntry in Bukkit.getBanList<BanList<*>>(BanList.Type.IP).banEntries) {
                if (target == banEntry.target) return true
            }
            return false
        }

        fun ban(target: String, reason: String?, expires: Date?, source: String?) {
            Bukkit.getScheduler().runTask(ActiveCraftCore.INSTANCE, Runnable {
                val event = PlayerIpBanEvent(target, true, reason, expires, source)
                Bukkit.getPluginManager().callEvent(event)
                if (!event.cancelled) {
                    Bukkit.getBanList<BanList<*>>(BanList.Type.IP).addBan(target, reason, expires, source)
                    for (player in Bukkit.getOnlinePlayers()) {
                        if (player.address.address.toString().replace("/", "") == target) {
                            val profile = Profile.of(player)
                            val messageSupplier: MessageSupplier =
                                profile.getMessageSupplier(ActiveCraftCore.INSTANCE)!!
                            profile.timesIpBanned += 1
                            Bukkit.getScheduler().runTask(ActiveCraftCore.INSTANCE, Runnable {
                                val expirationString = if (getRemainingAsString(expires).equals(
                                        "never",
                                        ignoreCase = true
                                    )
                                ) messageSupplier.getMessage("command.banscreen.expiration-permanent") else messageSupplier.getFormatted(
                                    "command.banscreen.expiration",
                                    MessageFormatter(
                                        messageSupplier.activeCraftMessage,
                                        "time",
                                        getRemainingAsString(expires)
                                    )
                                )
                                player.kickPlayer(
                                    messageSupplier.getMessage("command.banscreen.ip-title", ChatColor.RED)
                                            + "\n \n" + expirationString
                                            + "\n" + messageSupplier.getFormatted(
                                        "command.banscreen.reason",
                                        MessageFormatter(
                                            messageSupplier.activeCraftMessage,
                                            "reason",
                                            reason ?: messageSupplier.reasons.moderatorBanned
                                        )
                                    )
                                )
                            })
                        }
                    }
                }
            })
        }

        fun ban(target: Player, reason: String?, expires: Date?, source: String?) {
            ban(target.name, reason, expires, source)
        }

        fun unban(target: String?) {
            if (!isValidInet4Address(target)) {
                return
            }
            val entry = Bukkit.getBanList<BanList<*>>(BanList.Type.IP).getBanEntry(target!!) ?: return
            val event = PlayerIpBanEvent(target, false, entry.reason, entry.expiration, entry.source)
            Bukkit.getPluginManager().callEvent(event)
            if (event.cancelled) return
            Bukkit.getBanList<BanList<*>>(BanList.Type.IP).pardon(target)
        }

        fun unban(target: Player) {
            unban(target.name)
        }

        val bans: Set<BanEntry>
            get() = Bukkit.getBanList(BanList.Type.IP).banEntries

        fun getBanEntry(target: String?): BanEntry? {
            return Bukkit.getBanList<BanList<*>>(BanList.Type.IP).getBanEntry(target!!)
        }

        fun getBanEntry(target: Player): BanEntry? {
            return Bukkit.getBanList<BanList<*>>(BanList.Type.IP).getBanEntry(target.name)
        }
    }

    object Name {
        fun isBanned(target: String?): Boolean {
            return Bukkit.getBanList<BanList<*>>(BanList.Type.NAME).isBanned(target!!)
        }

        fun ban(target: String, reason: String?, expires: Date?, source: String?) {
            Bukkit.getScheduler().runTask(ActiveCraftCore.INSTANCE, Runnable {
                val event = PlayerBanEvent(target, reason, expires, source)
                Bukkit.getPluginManager().callEvent(event)
                if (!event.cancelled) {
                    Bukkit.getBanList<BanList<*>>(BanList.Type.NAME).addBan(target, reason, expires, source)
                    if (Bukkit.getPlayer(target) == null) return@Runnable
                    val profile = Profile.of(target) ?: return@Runnable
                    profile.timesBanned += 1
                    Bukkit.getScheduler().runTask(ActiveCraftCore.INSTANCE, Runnable {
                        val messageSupplier: MessageSupplier = profile.getMessageSupplier(ActiveCraftCore.INSTANCE)!!
                        val expirationString =
                            if (getRemainingAsString(expires).equals("never", ignoreCase = true)) {
                                messageSupplier.getMessage("command.banscreen.expiration-permanent")
                            } else {
                                messageSupplier.getFormatted(
                                    "command.banscreen.expiration",
                                    MessageFormatter(
                                        messageSupplier.activeCraftMessage,
                                        "time",
                                        getRemainingAsString(expires)
                                    )
                                )
                            }
                        Bukkit.getPlayer(target)!!.kickPlayer(
                            messageSupplier.getMessage("command.banscreen.title")
                                    + "\n \n" + expirationString
                                    + "\n" + messageSupplier.getFormatted(
                                "command.banscreen.reason",
                                MessageFormatter(
                                    messageSupplier.activeCraftMessage,
                                    "reason",
                                    reason ?: messageSupplier.reasons.moderatorBanned
                                )
                            )
                        )
                    })
                }
            })
        }

        fun ban(target: Player, reason: String?, expires: Date?, source: String?) {
            ban(target.name, reason, expires, source)
        }

        fun unban(target: String) {
            val event = PlayerUnbanEvent(target)
            Bukkit.getPluginManager().callEvent(event)
            if (event.cancelled) return
            Bukkit.getBanList<BanList<*>>(BanList.Type.NAME).pardon(target)
        }

        fun unban(target: Player) {
            unban(target.name)
        }

        val bans: Set<BanEntry>
            get() = Bukkit.getBanList<BanList<*>>(BanList.Type.NAME).banEntries

        fun getBanEntry(target: String?): BanEntry? {
            return Bukkit.getBanList<BanList<*>>(BanList.Type.NAME).getBanEntry(target!!)
        }

        fun getBanEntry(target: Player): BanEntry? {
            return Bukkit.getBanList<BanList<*>>(BanList.Type.NAME).getBanEntry(target.name)
        }
    }
}